TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+    // B) An array of valid type strings
+    // C) Invalid wrt the JSON Schema definition of the type keyword
+
+    [self pushPathComponent:kTWTJSONSchemaKeywordType];
+
+    static NSSet *allowedTypes = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        allowedTypes = [NSSet setWithObjects:kTWTJSONSchemaTypeKeywordAny, kTWTJSONSchemaTypeKeywordBoolean, kTWTJSONSchemaTypeKeywordNull, nil];
+    });
+
+    if ([allowedTypes containsObject:typeValue]) {
+        [self popPathComponent];
+        return [NSSet setWithObject:typeValue];
+    }
+
+    [self checkObjectIsArrayWithAtLeastOneItem:typeValue allowNil:NO];
+
+    [typeValue enumerateObjectsUsingBlock:^(NSString *typeString, NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [self checkObject:typeString isOneOf:[self validJSONTypeKeywords]];
+        [self popPathComponent];
+    }];
+
+    [self popPathComponent];
+    return [NSSet setWithArray:typeValue];
+}
+
+
+// Keywords: exclusiveMinimum & Maximum, uniqueItems
+- (BOOL)parseBooleanValueFromSchema:(NSDictionary *)schema forKey:(NSString *)key valueIfNotPresent:(BOOL)defaultValue
+{
+    [self pushPathComponent:key];
+    NSNumber *booleanObject = schema[key];
+    [self checkObject:booleanObject isOfClass:[NSNumber class] allowNil:YES];
+    [self popPathComponent];
+    return booleanObject ? [booleanObject boolValue] : defaultValue;
+}
+
+
+// Keywords: pattern
+- (NSString *)parseStringFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSString *string = schema[key];
+    [self checkObject:string isOfClass:[NSString class] allowNil:YES];
+    [self popPathComponent];
+    return string;
+}
+
+
+// Keywords: minimum, maximum
+- (NSNumber *)parseNumberFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSNumber *number = schema[key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+    [self popPathComponent];
+    return number;
+}
+
+
+// Keywords: multipleOf
+- (NSNumber *)parsePositiveNumberFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    NSNumber *number = schema[key];
+    if (!number) {
+        return nil;
+    }
+
+    [self pushPathComponent:key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+    if ([number doubleValue] <= 0) {
+        NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected number greater than zero.", key];
+        [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidValue reason:reason object:number];
+    }
+
+    [self popPathComponent];
+    return number;
+}
+
+
+// Keywords: min & maxLength, min & maxItems, min & maxProperties
+- (NSNumber *)parseUnsignedIntegerFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSNumber *number = schema[key];
```

Comment:
In the previous method, you get the number first and return if it’s non-`nil`. Any reason not to do that here?